warp/aggregator/aggregator_test.go

```diff
@@ -354,9 +378,72 @@ func TestAggregateSignatures(t *testing.T) {
 			expectedErr:     avalancheWarp.ErrInsufficientWeight,
 		},
 		{
-			name:        "early termination of signature fetching on passing threshold",
-			contextFunc: context.Background,
-			aggregatorFunc: func(ctrl *gomock.Controller) *Aggregator {
+			name: "context cancels halfway through signature fetching",
+			contextWithCancelFunc: func() (context.Context, context.CancelFunc) {
+				ctx, cancel := context.WithCancel(context.Background())
+				return ctx, cancel
+			},
+			aggregatorFunc: func(ctrl *gomock.Controller, cancel context.CancelFunc) *Aggregator {
+				state := validators.NewMockState(ctrl)
+				state.EXPECT().GetCurrentHeight(gomock.Any()).Return(pChainHeight, nil)
+				state.EXPECT().GetValidatorSet(gomock.Any(), gomock.Any(), gomock.Any()).Return(
+					vdrSet, nil,
+				)
+
+				client := NewMockSignatureGetter(ctrl)
+				client.EXPECT().GetSignature(gomock.Any(), nodeID1, gomock.Any()).DoAndReturn(
+					func(ctx context.Context, _ ids.NodeID, _ *avalancheWarp.UnsignedMessage) (*bls.Signature, error) {
+						select {
+						case <-ctx.Done():
+							err := ctx.Err()
+							require.ErrorIs(t, err, context.Canceled)
+							return nil, err
+						default:
+							// cancel the context and return the signature
+							cancel()
+							return sig1, nil
+						}
+					},
+				)
+				client.EXPECT().GetSignature(gomock.Any(), nodeID2, gomock.Any()).DoAndReturn(
+					func(ctx context.Context, _ ids.NodeID, _ *avalancheWarp.UnsignedMessage) (*bls.Signature, error) {
+						select {
+						case <-ctx.Done():
+							err := ctx.Err()
+							require.ErrorIs(t, err, context.Canceled)
+							return nil, err
+						default:
+							// We should not be able to get this signature since context cancelled in another go routine
```

Comment:
if this is the case then is it necessary to add this expected call to the mock?
if it's necessary, could we factor out the common code here (eg calling `client.EXPECT()` in a loop)

Reply:
I see, I can remove the default case then. I just thought it would be more clear that this case would NOT happen.